doc/source/guide/in_depth/functional_guide.rst

```diff
@@ -0,0 +1,257 @@
+
+The second assumption is made in order to guarantee that the optimization problem, in some way, makes sense.
+For example, if the field is not (totally) ordered (like for example the complex numbers) one cannot define optimization over it in a meaningful way.
+Or if the field does not have the least-upper-bound property, then attainment of an optimal solution cannot be guaranteed.
+See, for example, the Wikipedia articles on `field`_, `ordered field`_ and `least-upper-bound property`_.
+
+Note that it is possible to use the class without fulfilling these assumtptions, however in this case some of the mathematical results might not hold.
+
+.. _convex conjugate: https://en.wikipedia.org/wiki/Convex_conjugate
+.. _Legendre transformation: https://en.wikipedia.org/wiki/Legendre_transformation
+
+.. _field: https://en.wikipedia.org/wiki/Field_(mathematics)
+.. _ordered field: https://en.wikipedia.org/wiki/Ordered_field
+.. _least-upper-bound property: https://en.wikipedia.org/wiki/Least-upper-bound_property
+
+.. _implementation:
+
+Implementation of functionals
+=============================
+
+To define your own functional, you start by writing::
+
+    class MyFunctional(odl.solvers.Functional):
+        ...
+
+Since `Functional` is a subclass of `Operator`, it has the *abstract method* ``domain`` which need to be explicitly overridden by any subclass.
+
+``domain``: `Set`
+    The domain of this functional, i.e., the set of elements to which this functional can be applied.
+
+Moreover, there are several optional parameters associated with a functional. These are
+
+``linear``: `bool`, optional
+    If `True`, the functional is considered as linear.
+    Default: ``False``
+``smooth``: `bool`, optional
+    If `True`, assume that the functional is continuously differentiable.
+    Default: ``False``
+``concave``: `bool`, optional
+    If `True`, assume that the functional is concave.
+    Default: ``False``
+``convex``: `bool`, optional
+    If `True`, assume that the functional is convex.
+    Default: ``False``
+``grad_lipschitz``: `float`, optional
+    The Lipschitz constant of the gradient.
+    Default: infinity
+
+A functional also has a number of optional methods and properties associated with it.
+The default value of these are all to raise a `NotImplemetedError`.
+The properties are:
+
+ * ``gradient``. This returns the gradient operator of the functional, i.e., the operator that corresponds to the mapping
+
+   .. math::
+
+      x \to \nabla f(x),
+
+   where :math:`\nabla f(x)` is the element used to evaluate the (directional) derivative in a direction :math:`d` by :math:`\langle \nabla f(x), d \rangle`.
+
+ * ``conjugate_functional``. This is the convex conjugate functional, also known as the Legendre transform or Fenchel conjugate.
```

Comment:
Regarding the property name: I still think the "functional" part in the name is redundant, and that rather "convex" should be mentioned. We also don't repeat "operator" in `Operator.derivative`. 
My favorite is `convex_conj`. Other opinions, @adler-j, @hilding79, @ozanoktem?